### ISSUE TITLE
Add option to ignore Counter Reset 0 datapoint 

### ISSUE THREAD

**Issue by frogmaster** [OpenTSDB/opentsdb#460]
I have a  counter, that wraps once per day, which leaves me with graph having drops to zero once per day. It would be ok, if the datapoint is discarded instead or sent with null value instead. Handling the counter wrap value "0" client side is quite hard, because there is no distinction between real zero and wrap zero...

**Comment by kylebrandt** [comment]
I also wish was dropped. Having a zero can mess up reductions to a single number when using something like mean()

**Comment by luuse** [comment]
What kind of metrics do you have that are represented by counters (are they strictly increasing or can they decrease)? When you graph them do you still look at them as counters or as rates?

**Comment by frogmaster** [comment]
Mine specifically are interface octet counters, but from a rather nasty device, thus the daily wrap. They are strictly increasing, and i use rate.

**Comment by manolama** [comment]
Good idea, it's a little tricky to implement but doable.

**Comment by luuse** [comment]
I'm not entirely sure where you would like to drop the data point? Would you like it to be dropped at query time or not added to the database at all?

I haven't looked at the query code in a while and i'm still not sure if i understand how the graphs look currently. But i think to me the best solution would be to introduce an optional period parameter to the rate keyword. That should allow us to recognize when a counter rolls over as long as `v2 < v1` and continue the rate calculations as usual while still exposing real zeros?

**Comment by kylebrandt** [comment]
When a counter resets, it is impossible to now how much incremented. For
example observation A is 1000, and observation B is 14. We don't know much
the value increased after observation and the reset. Therefore we can't
take the derivative since we are missing that information.

So when doing a "rate" query, a more proper result for the derivative would
NaN. Since opentsdb doesn't support NaN the best thing would be to drop it.
This is because the resultant 0 calculation of the rate is lie (The rate
wasn't 0, OpenTSDB just can't calculate it).

The 0 behavior is particularly undesirable  when doing the reduction of a
time series generate by rate to a single number (i.e. avg() ) because it
throws the avg off.

**Comment by luuse** [comment]
So the problem you have is with counter resets and not how counter rollovers are currently handled? If so, does a properly configured `resetValue` correctly identify your resets? Finally, if `resetValue` identifies the resets, is it the `0` value that is currently returned when `resetValue` signals what you'd like to change?

**Comment by frogmaster** [comment]
Yes, that exactly what i need. insead of returning 0 value, return Null/Nan, or better yet, drop the particular datapoint from query results.

**Comment by oozie** [comment]
To illustrate the problem, rate of a monotonically increasing counter produces a low-quality data point immediately after counter reset (unless datapoint collection time happens at exactly the same time as counter reset, which is rather unlikely).
![counter_reset](https://cloud.githubusercontent.com/assets/1587140/6651241/6a5be684-c9f6-11e4-9d42-e8e1ffcd261c.png)
For every functional purpose, it's better to skip such datapoint on the rate metric.

**Comment by oozie** [comment]
I dug out #202 and I'm having a hard time deciding why fudging counter rates with Rate Ctr Max and Rate Ctr Reset was a good idea in the first place. The former tries to offset resets by a constant number and the latter defines a threshold at which to convert the rate to 0. It seems pretty attrocious, I can't think of a valid use case for this - what am I missing here?
Separately, the two text boxes take more than enough of UI's real estate.

**Comment by oozie** [comment]
@frogmaster @kylebrandt please give https://github.com/OpenTSDB/opentsdb/pull/465 a go.

**Comment by kylebrandt** [comment]
"I *think* David's idea for the reset value was as a kind of filter on anomalies. https://github.com/OpenTSDB/opentsdb/pull/202 For example, if you have a counter that is usually incremented by ~100 per time slice you could set "reset" to 150 and if the delta exceeded 150, (e.g. we go from 100 to 50000000 in one time slice) then we just reset to 0."

https://groups.google.com/forum/#!searchin/opentsdb/brandt$20counter/opentsdb/fw3LpYAiUl4/Etib1kT1lSsJ

In Bosun we store the Series type (provided by scollector) for most metrics. When that is available you get "auto" in the interface:

![image](https://cloud.githubusercontent.com/assets/1692624/6687661/56019596-cc83-11e4-88aa-f4fd2e6cee4a.png)

When you pick counter we just make it `counter{,,1}`:

![image](https://cloud.githubusercontent.com/assets/1692624/6687696/9d1e56ee-cc83-11e4-9848-263caeffc8c2.png)

So I look forward to my counter graphs not looking like this :-)

![image](https://cloud.githubusercontent.com/assets/1692624/6687742/0c282844-cc84-11e4-8495-c313bac0e935.png)

But time is tight for me so will be a while before I can test this :-/

**Comment by frogmaster** [comment]
oozie's pull request is fine, aside from one minor adjustment, which i commented on in pull request.

**Comment by manolama** [comment]
@kylebrandt Actually isn't it kind of good to have those drops to 0 there? It lets the user know that "hey, something happened here, a restart, a crash, etc". Especially in a monitoring system? ;)

But yeah, we can pull in #465 with some tweaks to allow the previous behavior and allow folks who really really want to hide the rollovers a drop flag.

And we'll be pushing up some code soon that will allow NaN's or Nulls on the output when DPs are missing and a downsampler is specified.

**Comment by kylebrandt** [comment]
That is really good news on NaN Null support. Would be nice to be able to
push those as well.

The drops to zero are not really good for the monitoring system:

   1. It throws off central tendency statistics, i.e. avg. This get thrown
   off more when you want a count of things that happened over time from a
   counter, because the avg rate (which is false) gets worse when multiplied
   by the time duration.
   2. Re: "Good to have the drops?" I don't think it is. If we I want
   either visually see or detect drops I can just ask for the counter data in
   its raw form. If point A > B I know it reset, generally easy to detect.
